Specify basic validation for federation membership endpoints

tulir opened the pull request:
<!-- Replace -->
Preview: https://pr2284--matrix-spec-previews.netlify.app
<!-- Replace -->

nexy7574 commented on data/api/server-server/invites-v1.yaml:
```diff
@@ -172,6 +172,17 @@ paths:
                   }
         "400":
           description: |-
+            The `M_INVALID_PARAM` error code is used to indicate one or more of the following:
```
Wouldn't `M_BAD_JSON` make sense?

> Request contained valid JSON, but it was malformed in some way, e.g. missing required keys, invalid values for keys.

richvdh reviewed (changes requested):
Thanks very much for taking this on. It's certainly a valuable improvement.

I'm a bit concerned that putting the validation requirements way down in the "error codes" section buries the lede a bit, and that people aren't going to notice it. I'd be more inclined to put it in the endpoint description.

Similarly, I'd phrase it as "Servers MUST validate the following: ..." rather than "[error code] is used in the following situations: ...".

tulir commented on data/api/server-server/invites-v1.yaml:
```diff
@@ -172,6 +172,17 @@ paths:
           description: |-
+            The `M_INVALID_PARAM` error code is used to indicate one or more of the following:
+
+            * The invite event fails a [signature check](/server-server-api/#validating-hashes-and-signatures-on-received-events).
```
I read it as those checks only applying to `invite_room_state`, which doesn't include the invite itself. The specific error code is fairly arbitrary (as mentioned in the other review thread)

richvdh commented:
> > I'm a bit concerned that putting the validation requirements way down in the "error codes" section buries the lede a bit, and that people aren't going to notice it. I'd be more inclined to put it in the endpoint description.
> > Similarly, I'd phrase it as "Servers MUST validate the following: ..." rather than "[error code] is used in the following situations: ...".
> 
> The main reason for both points was that invites already had `invite_room_state` validation in the response spec rather than the endpoint description. Should I move that stuff as well?

Ideally, I think so, yes, though at least in the case of `invite_room_state` we have a note on the request param that there is validation that must be done.

tulir commented:
Added the validation for GHSA-m5p2-vccg-8c9v, moved all of it from the response schemas to the request description and switched to use `M_MISSING_PARAM` for all errors

richvdh commented on data/api/server-server/invites-v2.yaml:
```diff
@@ -40,6 +40,30 @@ paths:
+        * One or more events fails a [signature check](/server-server-api/#validating-hashes-and-signatures-on-received-events).
+        * One or more events does not reside in the same room as the invite.
+          Note: Some room versions may require calculating the room ID for an
+          event rather than relying on the presence of `room_id`.
```
I feel like we've gone in the wrong direction here in terms of the error code. `M_MISSING_PARAM` seems like a terrible choice for param that isn't missing.

Can we update both halves to use something more sensible like `M_INVALID_PARAM`? Or just leave the precise error code unspecified here?

tulir replied:
If we want it to match implementations, probably just need to leave it unspecified. I guess it might be fine to specify something even if implementations don't match right now? Servers shouldn't be trying to catch and handle these errors anyway

richvdh commented on data/api/server-server/joins-v2.yaml:
```diff
@@ -247,6 +256,10 @@ paths:
             The error should be passed through to clients so that they
             may give better feedback to users.
 
+            If `M_MISSING_PARAM` is returned and the request is associated with a
```
The description for this endpoint specifies `M_INVALID_PARAM` for validation failures. Let's make sure they are consistent.

richvdh reviewed (approved):
lgtm. thanks.
